text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+with the mainline components to maintain this promise of
+compatibility.
+
+OTLP-Arrow supports conveying the gRPC metadta (i.e., http2 headers) using a dedicated `bytes` field.  Metadata is
+encoded using [hpack](https://datatracker.ietf.org/doc/rfc7541/) like a typical unary gRPC request.
+
+Specifically:
+
+#### OTLP/gRPC Receiver
+
+When Arrow is enabled, the OTLP receiver listens for both the standard unary gRPC service OTLP and OTLP-Arrow stream
+services.  Each stream uses an instance of the OTel-Arrow-Adapter's
+[Consumer](https://pkg.go.dev/github.com/f5/otel-arrow-adapter@v0.0.0-20230112224802-dafb6df21c97/pkg/otel/arrow_record#Consumer). Sets
+`client.Metadata` in the Context.
+
+#### OTLP/gRPC Exporter
+
+When Arrow is enabled, the OTLP exporter starts a fixed number of streams and repeatedly sends one `plog.Logs`,
+`ptrace.Traces`, or `pmetric.Metrics` item per stream request.  The `exporterhelper` callback first tries to get an
+available stream, blocking when none are available (or until the connection is downgraded), and then falls back to the
+standard unary gRPC path.  The stream-sending mechanism internally handles retries when failures are caused by streams
+restarting, while honoring the caller's context deadline, to avoid delays introduced by allowing these retries to go
+through the `exporterhelper` mechanism.
+
+Each stream uses an instance of the OTel-Arrow-Adapter's
+[Producer](https://pkg.go.dev/github.com/f5/otel-arrow-adapter@v0.0.0-20230112224802-dafb6df21c97/pkg/otel/arrow_record#Producer).
+
+When a stream fails specifically because the server does not recognize the Arrow service, it will not restart.  When all
+streams have failed in this manner, the connection downgrades by closing a channel, at which point the exporter behaves
+exactly as the standard OTLP exporter.
+
+The mechanism as described is vulnerable to partial failure scenarios.  When some of the streams are succeeding but some
+have failed with Arrow unsupported, the collector performance will be degraded because callers are blocked waiting for
+available streams.  The exact signal used to signal that Arrow and downgrade mechanism is seen as an area for future
+development.  [See the prototype's test for whether to downgrade.](https://github.com/open-telemetry/experimental-arrow-collector/blob/30e0ffb230d3d2f1ad9645ec54a90bbb7b9878c2/exporter/otlpexporter/internal/arrow/stream.go#L152)
+
+### Batch ID Generation
+
+The `batch_id` attribute is used by the message delivery mechanism. Each `BatchArrowRecords` issued must be associated with a
+unique `batch_id`. Uniqueness must be ensured in the scope of the stream opened by the call to the `ArrowStreamService`.
+This `batch_id` will be used in the `BatchStatus` object to acknowledge receipt and processing of the corresponding
+batch.
+A simple numeric counter can be used to implement this batch_id, the goal being to use the most concise id possible.
```

Comment:
Is there a reason not to make this a simple numeric counter always? What use case is there that doesn't work with a simple 64 bit integer sequence number (or even 32 bit and let it wrap)?

Reply:
There is no longer any good reason to keep this field a string so I will replace it with an int64.